### PR TITLE
Frequency scenario 3

### PR DESCRIPTION
**Pull request by matthewclegg**
This PR implements the third data generation scenario for frequency evaluation: publisher constant frequency.

The set_generatory.py file had grown to be rather large, so I have split the set generation classes into three files: a common set of base classes, classes for the cardinality estimation evaluation, and classes for frequency estimation evaluation.

In addition, I refactored the HomogeneousMultiSetGenerator so that it and the new PublisherConstantFrequencySetGenerator share a common base class.

### CODE REVIEW

File: src/simulations/tests/frequency_set_generator_test.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+          e = ExactMultiSet()
+          e.add_ids(s)
+          hists.append(LosslessEstimator()([e]))
+      self.assertLen(hists, 1)
+      self.assertEqual(hists[0], [2])
+
+  def test_homogeneous_pmf_multiset_generator_multiple_sets(self):
+      pmfgen = HomogeneousPmfMultiSetGenerator(
+          100, [2,1,2], [[1], [0,1], [0,0,1]], np.random.RandomState())
+      sets = [s for s in pmfgen]
```
Comment (my2457) [inline src/simulations/tests/frequency_set_generator_test.py:46]: not used?

File: src/simulations/tests/frequency_set_generator_test.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+          set_sizes=[1, 1],
+          freq_rates=[1, 1],
+          freq_cap=freq_cap,
+          random_state=np.random.RandomState())
+
+  def test_homogeneous_multiset_generator_factory_with_num_and_size(self):
+      f = HomogeneousMultiSetGenerator.get_generator_factory_with_num_and_size(
+          100, 3, 5, [1, 2, 3], 10)
+      gen = f(np.random.RandomState(1))
+      sets = [s for s in gen]
```
Comment (my2457) [inline src/simulations/tests/frequency_set_generator_test.py:123]: no need to have list comprehension here?

File: src/simulations/tests/frequency_set_generator_test.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+          100, 3, 3, 3)
+      gen = f(np.random.RandomState(1))
+      sets = [s for s in gen]
+      self.assertLen(sets, 3)
+
+  def test_publisher_constant_frequency_factory_with_set_size_list(self):
+      f = PublisherConstantFrequencySetGenerator.get_generator_factory_with_set_size_list(
+          100, [1, 2, 3], 3)
+      gen = f(np.random.RandomState(1))
+      sets = [s for s in gen]
```
Comment (my2457) [inline src/simulations/tests/frequency_set_generator_test.py:190]: same for `sets`?

File: src/simulations/tests/frequency_set_generator_test.py (hunk 1 above)
Comment (matthewclegg) [reply src/simulations/tests/frequency_set_generator_test.py:46]: Thanks, you're a very sharp-eyed reviewer!

File: src/simulations/tests/frequency_set_generator_test.py (hunk 2 above)
Comment (matthewclegg) [reply src/simulations/tests/frequency_set_generator_test.py]: Let me know if you like this better.

File: src/simulations/tests/frequency_set_generator_test.py (hunk 3 above)
Comment (matthewclegg) [reply src/simulations/tests/frequency_set_generator_test.py]: Ditto

File: src/simulations/frequency_set_generator.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        non-negative freq_rate parameter of the shifted-Possion distribution.
+      random_state: a numpy random state.
+      freq_cap: A positive integer which represents the maximum number
+        of times an ID can be seen in the returned set. 
+
+    Raises: AssertionError when (1) set_size_list and freq_rate_list do not
+      have equal length, (2) elements of freq_rate_list are not all
+      non-negative, or (3) freq_cap is not None or positive.
+    """
+
```
Comment (kungfucraig) [inline src/simulations/frequency_set_generator.py:169]: nit: delete newline

File: src/simulations/tests/frequency_set_generator_test.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+from wfa_cardinality_estimation_evaluation_framework.simulations.frequency_set_generator import HomogeneousPmfMultiSetGenerator
+from wfa_cardinality_estimation_evaluation_framework.simulations.frequency_set_generator import HomogeneousMultiSetGenerator
+from wfa_cardinality_estimation_evaluation_framework.simulations.frequency_set_generator import HeterogeneousMultiSetGenerator
+from wfa_cardinality_estimation_evaluation_framework.simulations.frequency_set_generator import PublisherConstantFrequencySetGenerator
+
+class FrequencySetGeneratorTest(parameterized.TestCase):
+
+  def test_homogeneous_pmf_multiset_generator_single_set(self):
+      pmfgen = HomogeneousPmfMultiSetGenerator(
+          100, [2], [[1]], np.random.RandomState())
```
Comment (kungfucraig) [inline src/simulations/tests/frequency_set_generator_test.py:34]: I think you probably want a constant seed on the RandomStates, here and below. If for no other reason that in principle having deterministic tests is good.

**Pull request merged**